File: backend/app/routes/vendedores.py
from fastapi import APIRouter, Depends, HTTPException
from passlib.context import CryptContext
from app.db import get_supabase
from app.auth import get_current_user

router = APIRouter()
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

FIELDS = "id, name, phone, role, is_active, balance, commission_rate, cidade_id, gerente_id, created_at"

@router.get("/")
async def list_vendedores(user=Depends(get_current_user)):
    q = get_supabase().table("vendedores").select(FIELDS).order("name")
    if user["role"] == "gerente":
        q = q.eq("gerente_id", user["id"])
    return q.execute().data

@router.get("/{vendedor_id}")
async def get_vendedor(vendedor_id: int, user=Depends(get_current_user)):
    res = get_supabase().table("vendedores").select(FIELDS).eq("id", vendedor_id).single().execute()
    if not res.data:
        raise HTTPException(status_code=404, detail="Vendedor não encontrado")
    return res.data

@router.post("/")
async def create_vendedor(body: dict, user=Depends(get_current_user)):
    name = body.get("name"); phone = body.get("phone"); password = body.get("password")
    if not all([name, phone, password]):
        raise HTTPException(status_code=400, detail="Dados incompletos")
    row = {
        "name": name, "phone": phone,
        "password_hash": pwd_context.hash(password),
        "role": body.get("role", "cambista"),
        "commission_rate": body.get("commission_rate"),
        "cidade_id": body.get("cidade_id"),
        "gerente_id": body.get("gerente_id"),
        "is_active": True, "balance": 0
    }
    res = get_supabase().table("vendedores").insert(row).execute()
    return res.data[0] if res.data else {}

@router.patch("/{vendedor_id}")
async def update_vendedor(vendedor_id: int, body: dict, user=Depends(get_current_user)):
    updates = dict(body)
    if "password" in updates:
        updates["password_hash"] = pwd_context.hash(updates.pop("password"))
    updates.pop("id", None)
    res = get_supabase().table("vendedores").update(updates).eq("id", vendedor_id).execute()
    return res.data[0] if res.data else {}

@router.delete("/{vendedor_id}")
async def delete_vendedor(vendedor_id: int, user=Depends(get_current_user)):
    get_supabase().table("vendedores").update({"is_active": False}).eq("id", vendedor_id).execute()
    return {"ok": True}

@router.post("/{vendedor_id}/change-password")
async def change_password(vendedor_id: int, body: dict, user=Depends(get_current_user)):
    if user["role"] != "admin" and user["id"] != vendedor_id:
        raise HTTPException(status_code=403, detail="Sem permissão")
    nova_senha = body.get("nova_senha")
    if not nova_senha:
        raise HTTPException(status_code=400, detail="Nova senha obrigatória")
    res = get_supabase().table("vendedores").select("password_hash").eq("id", vendedor_id).limit(1).execute()
    if not res.data:
        raise HTTPException(status_code=404, detail="Usuário não encontrado")
    if body.get("senha_atual"):
        if not pwd_context.verify(body["senha_atual"], res.data[0]["password_hash"]):
            raise HTTPException(status_code=401, detail="Senha atual incorreta")
    get_supabase().table("vendedores").update({"password_hash": pwd_context.hash(nova_senha)}).eq("id", vendedor_id).execute()
    return {"ok": True}

File: backend/src/routes/vendedores.js
const router = require('express').Router();
const auth = require('../middleware/auth');
const bcrypt = require('bcryptjs');
const supabase = require('../db');

// GET /vendedores
router.get('/', auth, async (req, res) => {
  try {
    let q = supabase.from('vendedores')
      .select('id, name, phone, role, is_active, balance, commission_rate, cidade_id, gerente_id, created_at')
      .order('name');
    // Gerente só vê seus cambistas
    if (req.user.role === 'gerente') q = q.eq('gerente_id', req.user.id);
    const { data, error } = await q;
    if (error) return res.status(400).json({ error: error.message });
    res.json(data);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

// GET /vendedores/:id
router.get('/:id', auth, async (req, res) => {
  try {
    const { data, error } = await supabase.from('vendedores')
      .select('id, name, phone, role, is_active, balance, commission_rate, cidade_id, gerente_id')
      .eq('id', req.params.id).single();
    if (error) return res.status(404).json({ error: 'Vendedor não encontrado' });
    res.json(data);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

// POST /vendedores
router.post('/', auth, async (req, res) => {
  try {
    const { name, phone, password, role, commission_rate, cidade_id, gerente_id } = req.body;
    if (!name || !phone || !password) return res.status(400).json({ error: 'Dados incompletos' });
    const password_hash = await bcrypt.hash(password, 10);
    const { data, error } = await supabase.from('vendedores')
      .insert({ name, phone, password_hash, role: role || 'cambista', commission_rate, cidade_id, gerente_id, is_active: true, balance: 0 })
      .select('id, name, phone, role, is_active').single();
    if (error) return res.status(400).json({ error: error.message });
    res.status(201).json(data);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

// PATCH /vendedores/:id
router.patch('/:id', auth, async (req, res) => {
  try {
    const updates = { ...req.body };
    if (updates.password) {
      updates.password_hash = await bcrypt.hash(updates.password, 10);
      delete updates.password;
    }
    delete updates.id;
    const { data, error } = await supabase.from('vendedores').update(updates).eq('id', req.params.id)
      .select('id, name, phone, role, is_active, balance, commission_rate').single();
    if (error) return res.status(400).json({ error: error.message });
    res.json(data);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

// DELETE /vendedores/:id (desativar)
router.delete('/:id', auth, async (req, res) => {
  try {
    const { error } = await supabase.from('vendedores').update({ is_active: false }).eq('id', req.params.id);
    if (error) return res.status(400).json({ error: error.message });
    res.json({ ok: true });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

module.exports = router;

// POST /vendedores/:id/change-password
router.post('/:id/change-password', auth, async (req, res) => {
  try {
    const id = parseInt(req.params.id);
    // Só pode trocar a própria senha (ou admin)
    if (req.user.role !== 'admin' && req.user.id !== id)
      return res.status(403).json({ error: 'Sem permissão' });

    const { senha_atual, nova_senha } = req.body;
    if (!nova_senha) return res.status(400).json({ error: 'Nova senha obrigatória' });

    const { data: users } = await supabase.from('vendedores')
      .select('password_hash').eq('id', id).limit(1);
    if (!users || !users.length) return res.status(404).json({ error: 'Usuário não encontrado' });

    if (senha_atual) {
      const ok = await bcrypt.compare(senha_atual, users[0].password_hash);
      if (!ok) return res.status(401).json({ error: 'Senha atual incorreta' });
    }

    const password_hash = await bcrypt.hash(nova_senha, 10);
    const { error } = await supabase.from('vendedores').update({ password_hash }).eq('id', id);
    if (error) return res.status(400).json({ error: error.message });
    res.json({ ok: true });
  } catch (e) { res.status(500).json({ error: e.message }); }
});
